src/tests/posix/sem_test.c

```diff
@@ -0,0 +1,184 @@
+}
+
+TEST_CASE("Test sem_wait") {
+    sem_t semaphore;
+    int value = -1;
+
+    test_assert_equal(sem_init(&semaphore, 0, SEMAPHORE_INIT_VALUE), ENOERR);
+    test_assert_equal(value, -1);
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, SEMAPHORE_INIT_VALUE); 	/* Test before sem_wait */
+
+    test_assert_equal(sem_wait(&semaphore), ENOERR);
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, SEMAPHORE_INIT_VALUE - 1); /* Test after sem_wait */
+}
+
+TEST_CASE("Test sem_post") {
+    sem_t semaphore;
+    int value = -1;
+
+    test_assert_equal(sem_init(&semaphore, 0, SEMAPHORE_INIT_VALUE), ENOERR);
+    test_assert_equal(value, -1);
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, SEMAPHORE_INIT_VALUE); 	/* Test before sem_wait */
+
+    test_assert_equal(sem_wait(&semaphore), ENOERR);
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, SEMAPHORE_INIT_VALUE - 1); /* Test after sem_wait */
+
+    test_assert_equal(sem_post(&semaphore), ENOERR);
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, SEMAPHORE_INIT_VALUE); 	/* Test after sem_post */
+}
+
+TEST_CASE("Test sem_trywait") {
+    sem_t semaphore;
+    int value = -1;
+
+    test_assert_equal(sem_init(&semaphore, 0, 1), ENOERR);
+    test_assert_equal(value, -1);
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, 1);
+
+    test_assert_equal(sem_trywait(&semaphore), ENOERR);
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, 0);
+
+    test_assert_equal(sem_trywait(&semaphore), -EAGAIN);
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, 0);
+}
+
+TEST_CASE("Test sem_timedwait") {
+    sem_t semaphore;
+    struct timespec now, ts;
+    int value = -1;
+
+    test_assert_equal(sem_init(&semaphore, 0, 1), ENOERR);
+    test_assert_equal(value, -1);
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, 1);
+
+    clock_gettime(CLOCK_REALTIME, &ts);
+    ts.tv_nsec += 100000000;		/* Wait 0.1s */
+    test_assert_equal(sem_timedwait(&semaphore, &ts), ENOERR);
+    clock_gettime(CLOCK_REALTIME, &now);
+    test_assert(ts_cmp(&ts, &now, >));	/* Assert ts is after now */
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, 0);
+
+    clock_gettime(CLOCK_REALTIME, &ts);
+    ts.tv_nsec += 100000000;		/* Wait 0.1s */
+    test_assert_equal(sem_timedwait(&semaphore, &ts), -ETIMEDOUT);
+    clock_gettime(CLOCK_REALTIME, &now);
+    test_assert(ts_cmp(&ts, &now, <=));
+    test_assert_equal(sem_getvalue(&semaphore, &value), ENOERR);
+    test_assert_equal(value, 0);
+}
+
+static pthread_t sem_post_thread;
+static sem_t semaphore_test_timedwait;
```

Comment:
You can declare it as a local variable and pass pointer to it to `sem_post_run(void *arg) ` as arg through `pthread_create(&sem_post_thread, NULL, sem_post_run, &semaphore_test_timedwait)`